### ISSUE TITLE
Add Paillier CUDA Support

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/PySyft#2747]
**Is your feature request related to a problem? Please describe.**
We want to be able to run Paillier encryption on CUDA (encrypted prediction of linear models as well as secure aggregation for federated learning)

**Describe the solution you'd like**

Wrap this library - https://github.com/data61/cuda-fixnum
Port this library to PyTorch - https://python-paillier.readthedocs.io/en/develop/

**Describe alternatives you've considered**
Wrap https://python-paillier.readthedocs.io/en/develop/

**Additional context**
I have good reason to believe that the CUDA library is the fastest CUDA implementation of paillier out there.

**Comment by PlamenHristov** [comment]
Maybe I can give it a try ?

**Comment by abogaziah** [comment]
@PlamenHristov how far have you reached? maybe we can work together on it!

**Comment by PlamenHristov** [comment]
Hey @abogaziah,

I've gotten to a place, where I've setup an env for extending PyTorch with a CUDA implementation of as I believe the best way forward is to "Port this library to PyTorch - https://python-paillier.readthedocs.io/en/develop/". 
I don't think "Wrap this library - https://github.com/data61/cuda-fixnum" is necessary ? @iamtrask  correct me if I am wrong.

I could send you a link to my forked repo and we can work there?

**Comment by abogaziah** [comment]
sounds great!

**Comment by iamtrask** [comment]
@PlamenHristov and @abogaziah, I 100% support porting python-paillier to PyTorch. there will be challenges (primarily with representing large numbrees, which in theory CRTTensor can do but it will be tricky). The best part about doing it this way is that it means your tensor will automatically work in JS and iOS/Android by default.

Also, the author of python-paillier is a great guy and is often around the project (@hardbyte). He may be willing to consult.

Also, if you're no yet a member of a development team, you should 100% apply to one with this project. This would help provide you support as you go from the broader crypto/PySyft community here.

**Comment by unzvfu** [comment]
Hi all, author of cuda-fixnum here; @hardbyte sent me your way.

I started working on a Python wrapper for cuda-fixnum last year, and it's still a goal of the project. Unfortunately I don't have time to work on the HLL interface part of the project at the moment, a situation which is not likely to change in the near future. The current state of my Python work is, frankly, a _mess_, and probably not worth anyone's time. But if a brave soul is willing to take up the mantle, then feel free to get in contact and I'll do my best to guide you away from the pitfalls I came across.
